src/d3a/d3a_core/redis_connections/aggregator_connection.py

```diff
@@ -49,6 +52,91 @@ def add_batch_trade_event(self, device_uuid, event):
     def add_batch_finished_event(self, device_uuid, event):
         self._add_batch_event(device_uuid, event, self.batch_finished_events)
 
+    def crud_aggregator_callback(self, payload):
+        message = json.loads(payload["data"])
+        if message["type"] == "CREATE":
+            self._create_aggregator(message)
+        elif message["type"] == "DELETE":
+            self._delete_aggregator(message)
+        elif message["type"] == "SELECT":
+            self._select_aggregator(message)
+
+    def _select_aggregator(self, message):
+        if message['aggregator_uuid'] not in self.selected_aggregators:
+            msg = f"{message['aggregator_uuid']} aggregator not found."
+            error_response_message = {
+                "status": "error", "aggregator_uuid": message['aggregator_uuid'],
+                "device_uuid": message['device_uuid'],
+                "transaction_id": message['transaction_id'],
+                "msg": msg
+            }
+            self.redis_db.publish(
+                "crud_aggregator_response", json.dumps(error_response_message)
+            )
+        elif message['device_uuid'] in self.device_aggregator_mapping:
+            msg = f"Device already have selected " \
+                  f"{self.device_aggregator_mapping[message['device_uuid']]}"
+            error_response_message = {
+                "status": "error", "aggregator_uuid": message['aggregator_uuid'],
+                "device_uuid": message['device_uuid'],
+                "transaction_id": message['transaction_id'],
+                "msg": msg
+            }
+            self.redis_db.publish(
+                "crud_aggregator_response", json.dumps(error_response_message)
+            )
+        else:
+            self.aggregator_device_mapping[message['aggregator_uuid']].\
+                append(message['device_uuid'])
+            self.device_aggregator_mapping[message['device_uuid']] = message['aggregator_uuid']
+            success_response_message = {
+                "status": "SELECTED", "aggregator_uuid": message['aggregator_uuid'],
+                "device_uuid": message['device_uuid'],
+                "transaction_id": message['transaction_id']}
+            self.redis_db.publish(
+                "crud_aggregator_response", json.dumps(success_response_message)
+            )
+
+    def _create_aggregator(self, message):
+        if message['transaction_id'] not in self.selected_aggregators:
+            with self.lock:
+                self.selected_aggregators.append(message['transaction_id'])
```

Comment:
why do you need `selected_aggregators`? Isnt `aggregator_device_mapping.keys()` sufficient?

Reply:
You are right, will remove it!